File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    bool sContainsT(unordered_map<char, int>& s_mp, unordered_map<char, int>& t_mp) {
        for (auto& a : t_mp) {
            if (s_mp.find(a.first) == s_mp.end()) return false;
            if (s_mp[a.first] < a.second) return false;
        }
        return true;
    }
    void eraseFromMap(unordered_map<char, int>& s_mp, char key) {
        if (s_mp[key] > 1) s_mp[key]--;
        else s_mp.erase(key);
    } 
    string minWindow(string s, string t) {
        int m = s.size(); 
        int n = t.size();
        unordered_map<char, int> s_mp, t_mp;
        
        int left = 0;
        int right = 0;
        
        for (int i = 0; i < n; ++i) {
            t_mp[t[i]]++;
        }
        int res = INT_MAX;
        int index = -1;
        while (right < m) {
            s_mp[s[right]]++;
            while (left <= right && sContainsT(s_mp, t_mp)) {

                if (res > right - left + 1) {
                    index = left;
                    res = right - left + 1;
                }              
                eraseFromMap(s_mp, s[left]);
                left++;
            }
            right++;
        }
        return res == INT_MAX ? "" : s.substr(index, res);
    }
};

File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public boolean sContainsT(Map<Character, Integer> sMap, Map<Character, Integer> tMap) {
        for (Map.Entry<Character, Integer> entry: tMap.entrySet()) {
            char ch = entry.getKey();
            int val = entry.getValue();
            if (!sMap.containsKey(ch) || sMap.get(ch) < val) return false;
        }
        return true;
    }
    public void eraseFromMap(Map<Character, Integer> map, char ch) {
        if (map.get(ch) == 1) map.remove(ch);
        else map.put(ch, map.get(ch) - 1);
    }
    public String minWindow(String s, String t) {
        int m = s.length();
        int n = t.length();
        int left = 0;
        int right = 0;
        int index = -1;
        int resLen = Integer.MAX_VALUE;
        Map<Character, Integer> sMap = new HashMap<>();
        Map<Character, Integer> tMap = new HashMap<>();
        for (int i = 0; i < n; ++i) {
            char ch = t.charAt(i);
            tMap.put(ch, tMap.getOrDefault(ch, 0) + 1);
        }
        while (right < m) {
            char ch = s.charAt(right);
            sMap.put(ch, sMap.getOrDefault(ch, 0) + 1);

            while (sContainsT(sMap, tMap)) {
                eraseFromMap(sMap, s.charAt(left));
                // Capture a res
                if (resLen > right - left + 1) {
                    resLen = right - left + 1;
                    index = left;
                }
                left++;
            }
            right++;
        } 
        if (resLen == Integer.MAX_VALUE) return "";
        return s.substring(index, index + resLen);
    }
}